cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/SystemArtifactConfig.java

```diff
@@ -0,0 +1,324 @@
+        this.parents.add(parentRange);
+      }
+      return this;
+    }
+
+    SystemArtifactConfig build() throws InvalidArtifactException {
+      validateParentList(parents);
+      validatePluginList(plugins);
+      SystemArtifactConfig config = new SystemArtifactConfig(artifactId, artifactFile,
+        Collections.unmodifiableSet(parents), Collections.unmodifiableSet(plugins));
+      return config;
+    }
+  }
+
+  /**
+   * Deserializer for ArtifactRange in a SystemArtifactConfig. Artifact ranges are expected to be able to be
+   * parsed via {@link ArtifactRange#parse(String)}. Currently, system artifacts can only extend other system artifacts.
+   */
+  private static final class ArtifactRangeDeserializer
+    implements JsonDeserializer<ArtifactRange>, JsonSerializer<ArtifactRange> {
+
+    @Override
+    public ArtifactRange deserialize(JsonElement json, Type typeOfT,
+                                     JsonDeserializationContext context) throws JsonParseException {
+      if (!json.isJsonPrimitive()) {
+        throw new JsonParseException("Parent artifacts should be strings.");
+      }
+
+      try {
+        return ArtifactRange.parse(Id.Namespace.SYSTEM, json.getAsString());
+      } catch (InvalidArtifactRangeException e) {
+        throw new JsonParseException(e.getMessage(), e);
+      }
+    }
+
+    @Override
+    public JsonElement serialize(ArtifactRange src, Type typeOfSrc, JsonSerializationContext context) {
+      return new JsonPrimitive(src.toNonNamespacedString());
+    }
+  }
+
+  /**
+   * Deserializer for SystemArtifactConfig. Used to check validity of an artifact config. For example,
+   * a plugin class should not appear more than once in the list of plugins, and a parent artifact should not
+   * appear more than once in the list of parents.
+   */
+  private static final class SystemArtifactConfigDeserializer
+    implements JsonDeserializer<SystemArtifactConfig>, JsonSerializer<SystemArtifactConfig> {
+    private static final Type PLUGINS_TYPE = new TypeToken<Set<PluginClass>>() { }.getType();
+    private static final Type PARENTS_TYPE = new TypeToken<Set<ArtifactRange>>() { }.getType();
+
+    @Override
+    public SystemArtifactConfig deserialize(JsonElement json, Type typeOfT,
+                                            JsonDeserializationContext context) throws JsonParseException {
+      if (!json.isJsonObject()) {
+        throw new JsonParseException("Config file must be a JSON Object.");
+      }
+      JsonObject obj = json.getAsJsonObject();
+
+      // deserialize fields
+      Set<ArtifactRange> parents = context.deserialize(obj.get("parents"), PARENTS_TYPE);
+      parents = parents == null ? Collections.<ArtifactRange>emptySet() : parents;
+      Set<PluginClass> plugins = context.deserialize(obj.get("plugins"), PLUGINS_TYPE);
+      plugins = plugins == null ? Collections.<PluginClass>emptySet() : plugins;
+
+      try {
+        validateParentList(parents);
```

Comment:
The validation of parentList and pluginList needs to be performed for all artifacts isn't it? Is it specific to SystemArtifacts?